feature/policy_forwarding/encapsulation/otg_tests/encap_gre_ipv4/encap_gre_ipv4_test.go

```diff
@@ -0,0 +1,1062 @@
+	rule2.GetOrCreateIpv6().DestinationAddress = ygot.String(fmt.Sprintf("%s/32", otgPort3.IPv4))
+	rule2.GetOrCreateAction().SetNextHop(dutPort3.IPv6)
+
+	// Rule 3: Match IPV4-SRC2 and encapsulate to 32 IPv4 GRE destinations
+	rule3 := policy.GetOrCreateRule(3)
+	rule3.GetOrCreateIpv4().DestinationAddress = ygot.String(fmt.Sprintf("%s/128", otgPort2.IPv4))
+	encapGre3 := rule3.GetOrCreateAction().GetOrCreateEncapsulateGre()
+	for i, dest := range tunnelDestinations {
+		targetName := fmt.Sprintf("gre%d", i)
+		encapGre3.GetOrCreateTarget(targetName).Source = ygot.String(dutlo0Attrs.IPv4)
+		encapGre3.GetOrCreateTarget(targetName).Destination = ygot.String(dest)
+	}
+
+	// Rule 4: Match IPV6-SRC2 and encapsulate to 32 IPv4 GRE destinations
+	rule4 := policy.GetOrCreateRule(4)
+	rule4.GetOrCreateIpv6().DestinationAddress = ygot.String(fmt.Sprintf("%s/128", otgPort3.IPv6))
+	encapGre4 := rule4.GetOrCreateAction().GetOrCreateEncapsulateGre()
+	for i, dest := range tunnelDestinations {
+		targetName := fmt.Sprintf("gre%d", i)
+		encapGre4.GetOrCreateTarget(targetName).Source = ygot.String(dutlo0Attrs.IPv4)
+		encapGre4.GetOrCreateTarget(targetName).Destination = ygot.String(dest)
+	}
+
+	// Apply the policy to DUT Port 1
+	ni.GetOrCreatePolicyForwarding().GetOrCreateInterface(interfaceName).ApplyForwardingPolicy = ygot.String(policyName)
+	gnmi.Replace(t, dut, gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).PolicyForwarding().Config(), ni.PolicyForwarding)
+}
+
+// ConfigureQoSClassifier configures QoS classifier for incoming traffic on ATE Port1.
+func configureQoSClassifier(t *testing.T, dut *ondatra.DUTDevice) {
+	interfaceName := dut.Port(t, "port1").Name()
+	classifierName := "qos-classifier-1"
+	ipv4DscpValues := []uint8{0, 8, 16, 24, 32, 40, 48, 56}
+	ipv6DscpValues := []uint8{0, 32, 64, 96, 128, 160, 192, 224}
+	if deviations.QosClassifierDscpRemarkOcUnsupported(dut) {
+		t.Logf("Configuring qos through CLI")
+		configureQoSClassifierFromCLI(t, dut, classifierName, interfaceName, ipv4DscpValues, ipv6DscpValues)
+	} else {
+		t.Logf("Configuring qos through OC")
+		configureQoSClassifierFromOC(t, dut, classifierName, interfaceName, ipv4DscpValues, ipv6DscpValues)
+	}
+}
+
+func configureQoSClassifierFromCLI(t *testing.T, dut *ondatra.DUTDevice, classifierName string, interfaceName string, ipv4DscpValues []uint8, ipv6DscpValues []uint8) {
+	gnmiClient := dut.RawAPIs().GNMI(t)
+	qosConfig := qosClassifierCliConfig(dut, classifierName, interfaceName, ipv4DscpValues, ipv6DscpValues)
+	t.Logf("Push the CLI Qos config:%s", dut.Vendor())
+	gpbSetRequest := buildCliSetRequest(qosConfig)
+	if _, err := gnmiClient.Set(context.Background(), gpbSetRequest); err != nil {
+		t.Errorf("Failed to set qos classifier from cli: %v", err)
+	}
+}
+
+func configureQoSClassifierFromOC(t *testing.T, dut *ondatra.DUTDevice, classifierName string, interfaceName string, ipv4DscpValues []uint8, ipv6DscpValues []uint8) {
```

Comment:
use internal/cfgplugins/qos.go.

Note that https://github.com/openconfig/featureprofiles/pull/4218 @crc-kt is working on the same thing, so you should coordinate with them.